src/filter.ts

```diff
@@ -32,12 +32,18 @@ export async function filter(input: any, predicate: any): Promise<any[]> {
   }
 
   // tslint:disable-next-line:no-any
-  const output: any[] = [];
+  const predicateResults: any = Array.isArray(input) ? new Array(input.length) : {};
   // tslint:disable-next-line:no-any
   await map(input, async (value: any, key: any) => {
```

Comment:
this ordering piece is already handled by map. You can just do `const results = await map(...)`

Reply:
ah i chose to do it this way bc `map` always returns an array, which is less convenient to use when `input` is an object

Reply:
I'm not sure I understand why that matters if you're always returning an array regardless.

Wouldn't this piece still work exactly the same way?
```ts
  // tslint:disable-next-line:no-any
  const output: any[] = [];
  for (const k in input) {
    if (predicateResults[k]) {
      output.push(input[k]);
    }
  }
```

Reply:
`k` is either an index or an object key depending on the input, so we'd essentially be relying on key ordering if we use `const predicateResults: any[] = await map(...)`

Reply:
🤦 was thinking about this as a variant of map and not as returning original inputs. This makes sense now.